02-spatial-data.Rmd

```diff
@@ -670,6 +670,115 @@ st_crs(points_sfc_wgs) # print CRS (only first 4 lines of output shown)
 #> ...
 ```
 
+### The sfheaders package
+
+```{r sfheaers-setup, echo=FALSE}
+## Detatch {sf} to remove 'print' methods
+## because I want to show the underlying structure 
+##
+## library(sf) will be called later
+# unloadNamespace("sf") # errors
+# pkgload::unload("sf")
+```
+
+**sfheaders** is an R package that speeds-up the construction of `sf` objects, with a focus on:
+
+1. **building** **sf** objects from vectors, matrices and data frames, rapidly, and without depending on the **sf** library
+2. **exposing** its underlying C++ code through header files (hence the name, **sfheaders**), enabling others to extend it using compiled and fast-running code
+
+The intention is to allow every **sfheaders** operation implemented in R to have a C++ implementation, as described in the [`Cpp` vignette](https://dcooley.github.io/sfheaders/articles/Cpp.html).
+For most people users the R functions will be more than sufficient to benefit from the computational speed of the package.
+**sfheaders** was developed separately from **sf**, but aims to be fully compatible, creating valid `sf` objects of the type described in preceding sections.
+
+The simplest use-case for `**sfheaders**` is demonstrated in the code chunks below with examples of building  `sfg`, `sfc` and `sf` objects showing:
+
+- a vector converted to `sfg_POINT`
+- a matrix converted to `sfg_LINESTRING`
+- a data.frame converted to `sfg_POLYGON`
+
+We will start by creating the simplest possible `sfg` object, a single coordinate pair, assigned to a vector named `v`:
+
+```{r}
+v = c(1, 1)
+v_sfg_sfh = sfheaders::sfg_point(obj = v)
+```
+
+```{r sfheaders-sfg_point, eval=FALSE} 
+v_sfg_sfh # printing without sf loaded
+#>      [,1] [,2]
+#> [1,]    1    1
+#> attr(,"class")
+#> [1] "XY"    "POINT" "sfg" 
+```
+
+```{r, eval=FALSE, echo=FALSE}
+v_sfg_sfh = sf::st_point(v)
+```
+
+The example above shows how the `sfg` object `v_sfg_sfh` is printed when **sf** is not loaded, demonstrating its underlying structure.
+When **sf** is loaded (as is the case here), the result of the above command is indistinguishable from `sf` objects:
+
+```{r}
+v_sfg_sf = st_point(v, c(1, 2))
+print(v_sfg_sf) == print(v_sfg_sfh)
+```
+
+```{r, echo=FALSE, eval=FALSE}
+# (although `sfg` objects created with **sfheaders** have a dimension while `sfg` objects created with the **sf** package do not)
+waldo::compare(v_sfg_sf, v_sfg_sfh)
+dim(v_sfg_sf)
+dim(v_sfg_sfh)
+attr(v_sfg_sfh, "dim")
+```
+
+The next examples shows how **sfheaders** creates objects creates `sfg` objects from matrices:
```

Comment:
delete first instance of creates objects

Reply:
:+1: